### PR TITLE
MultiHeadAttention Layer Bug Fix

### PR DESCRIPTION
**Pull request by kumarutkarsh1248**
This PR address issue: [3969](https://github.com/mlpack/mlpack/issues/3969)

Fixed the forward pass implementation in the MultiHeadAttention layer to produce the correct output.

Next Steps:
Get feedback on the changes and explore possible optimizations.
Once finalized, extend the fixes to the backward pass and gradient computation.

### CODE REVIEW

File: src/mlpack/methods/ann/layer/multihead_attention.hpp [hunk 1]
```diff
@@ -283,11 +283,12 @@ class MultiheadAttention : public Layer<MatType>
...
 
-  //! Two dimensional Attention Mask of shape (tgtSeqLen, srcSeqLen).  Takes
+  //! Two dimensional Attention Mask of shape (tgtSeqLen, srcSeqLen, batchSize).  Takes
   //! the values [-Inf, 0]
-  MatType attnMask;
+  CubeType attnMask;
 
-  //! Key Padding Mask.  Takes the values [-Inf, 0]
+  //! Key Padding Mask.  The shape of keyPaddingMask : (batchSize, srcSeqLen)
+  //! the values [-Inf, 0]
```
Comment (rcurtin) [inline src/mlpack/methods/ann/layer/multihead_attention.hpp]:
```suggestion
  //! Takes the values [-Inf, 0].
```
(Just preserving the previous comment.)

File: src/mlpack/methods/ann/layer/multihead_attention_impl.hpp [hunk 2]
```diff
@@ -132,28 +132,41 @@ Forward(const MatType& input, MatType& output)
...
   {
-    if (attnMask.n_rows != tgtSeqLen || attnMask.n_cols != srcSeqLen)
+    if (attnMask.n_slices != input.n_cols || attnMask.n_rows != tgtSeqLen || attnMask.n_cols != srcSeqLen)
       Log::Fatal << "The size of the 'attn_mask' is not correct.\n";
-    scores.each_slice() += attnMask;
+    for (size_t i = 0; i < batchSize; ++i)
+    {
+      for (size_t h = 0; h < numHeads; ++h)
+      {
+          scores.slice(i * numHeads + h) += attnMask.slice(i).t();
```
Comment (rcurtin) [inline src/mlpack/methods/ann/layer/multihead_attention_impl.hpp]: If we're always transposing the attention mask, can we just change its definition to be size `(srcSeqLen, tgtSeqLen, batchSize)` instead?

(Also this line should only be two-space indented, not four.)

File: src/mlpack/methods/ann/layer/multihead_attention_impl.hpp [hunk 3]
```diff
@@ -132,28 +132,41 @@ Forward(const MatType& input, MatType& output)
...
   {
-    if (keyPaddingMask.n_rows != 1 || keyPaddingMask.n_cols != srcSeqLen)
+    if (keyPaddingMask.n_rows != input.n_cols || keyPaddingMask.n_cols != srcSeqLen)
         Log::Fatal << "The size of the 'keyPaddingMask' is not correct.\n";
-    scores.each_slice() += repmat(keyPaddingMask, tgtSeqLen, 1);
+    for (size_t i = 0; i < batchSize; ++i)
+    {
+      for (size_t h = 0; h < numHeads; ++h)
+      {
+          scores.slice(i * numHeads + h) += repmat(keyPaddingMask.row(i), tgtSeqLen, 1).t();
```
Comment (rcurtin) [inline src/mlpack/methods/ann/layer/multihead_attention_impl.hpp]: Same here, can we change `keyPaddingMask` to have shape `(srcSeqLen, batchSize)` instead?

**Review by rcurtin — COMMENTED** [review]
Nice, this is looking good to me.  A couple questions:

 * Can you add a note to `HISTORY.md`?
 * Did you compile and run `mlpack_test` locally to test that the changes don't break anything?
 * Do these changes make inference work for mlpack-onnx in your local testing?

I just have a couple comments that could help improve the speed of the implementation.

**Comment by kumarutkarsh1248** [comment]
1. I will add a note for this in HISTORY.md in my next commit.

2. No, I didn’t test the change with mlpack_test because I first wanted to check if you were okay with the optimization or not. Now I think, apart from the transpose in padMask and attnMask, everything seems fine. So next, I will try to update backpropagation and gradient accordingly and make changes in the mask as you suggested.

3. Yes, I tested the change with mlpack-onnx, and I was able to get the correct inference.

File: src/mlpack/methods/ann/layer/multihead_attention_impl.hpp (hunk 2 above)
Comment (kumarutkarsh1248) [reply src/mlpack/methods/ann/layer/multihead_attention_impl.hpp]: Yes, I agree with this change. I will update it in the next commit

**Comment by rcurtin** [comment]
Awesome---sounds good on all three points. :+1:
